shacl12-node-expr/index.html

```diff
@@ -2256,7 +2256,7 @@ <h3>sh:expression</h3>
 				</div>
 				<p><em>The remainder of this section is informative.</em></p>
 				<p>
-					Note that the scope in the evaluation of expression constraints contains the variable <code>value</code>
+					Note that the scope in the evaluation of expression constraints contains the <a href="#VarExpression">variable</a> <code>value</code>
```

Comment:
At least:

```suggestion
					Note that the scope in the evaluation of expression constraints contains the <a href="#dfn-var-expression">variable</a> <code>value</code>
```
although this called a "var expression" not "a variable" so that seems to make it:

```suggestion
					Note that the scope in the evaluation of expression constraints contains the <a href="#VarExpression">var expression</a> <code>value</code>
```

The definition:

> A [blank node](https://www.w3.org/TR/rdf12-concepts/#dfn-blank-node) that is the [subject](https://www.w3.org/TR/rdf12-concepts/#dfn-subject) of the following properties is called a var expression with the [function name](https://www.w3.org/TR/shacl12-core/#dfn-function-name) shnex:VarExpression:

says the blank node _is_ the var expression and _is_ an RDF term.

If you want to keep it as the current design, please keep using "var expression" and then shacl12-rules can define `variable`. (There are only two other uses of "variable" in the node expressions doc at the moment.)

Reply:
We don't even need to talk VarExpressions at all here. The key is the term "scope". And scope is defined in SHACL 1.2 core as

					<li><code>scope</code> is a map from (key) <a>nodes</a> to individual (value) <a>nodes</a>.
					The empty map is written as <code>{}</code>.

So I think we can avoid the term variable in that definition and just refer to mappings in the scope, where the key in the map is the value of shnex:var of the var expression.

I guess the proper fix is then to create a deep link to the scope definition and reference that from node expressions.

Reply:
The map is from a string (taken from `shnex:var`) to RDF term.

If it is a node (RDF Term), then

```
[ shnex:var "foo" ] .
[ shnex:var "foo" ] .
```
is two different scope entries.

If there is no varExpression, how does a node expression work

```
  [ sparql:plus ( [ shnex:var foo" ] 1 ) ]
```
Something has to define 
```
   exprEval([ shnex:var "foo" ])
```
and that to be an error if "foo" is not in the scope, which is returning an empty list from scope lookup.

Reply:
I struggle to parse your response.

```
[ shnex:var "foo" ] .
[ shnex:var "foo" ] .
```

are two times the same expression. Why should those be "different scope entries"?

The sentence above the sparql:plus example seems unfinished. What do you mean?

We could change the handling of undefined variables but then how could we distinguish unbound (optional) parameters from those called by accident in a node expression? There is no "null" value here like in JavaScript.

Anyway, I made further adjustments to the PR and believe external documents can refer to the variable <code>scope</code> by name when they refer to evalExpr.

Reply:
> I struggle to parse your response.
> 
> ```
> [ shnex:var "foo" ] .
> [ shnex:var "foo" ] .
> ```
> 
> are two times the same expression.

And what _RDF term_ do they evaluate to? That's necessary because:

> scope is a map from (key) [nodes](https://www.w3.org/TR/rdf12-concepts/#dfn-rdf-node) to individual (value) [nodes](https://www.w3.org/TR/rdf12-concepts/#dfn-rdf-node).

(PS better to say "RDF term", not "nodes" which are the existing subject/objects of a graph)

Reply:
If "foo" is not in the scope, then this rule applies

otherwise evalExpr(expr, focusGraph, focusNode, scope) -> []

i.e. the var node expression returns the empty list.

And OK I have switched the definition to "term". If we want to be consistent and serious about that, I believe many many other places would be affected though, e.g. node expressions would need to be renamed to term expressions. This would probably alienate users though, as the term "term" is barely used outside of W3C specs, while "node" is well-known.

Reply:
This is probably me still shaking cobwebs off from being AFK a while, but I need a reminder on one of the less intuitive ideas from RDF:

Is a literal a node?
Is a literal a term?

Update for RDF 1.2:

Is a triple term a node?
Is a triple term a term?  (...If it isn't, why?)

Reply:
RDF Terms : https://www.w3.org/TR/rdf12-concepts/#section-terms 
IRIs, Blank nodes, literals, triple terms.

Nodes: https://www.w3.org/TR/rdf12-concepts/#dfn-node

"Nodes of a graph" -- the set of Subjects and Objects in use in a graph (not if only in the predicate position, of if only inside triple terms). As in "nodes and arcs" for graphs.

Triple terms: https://www.w3.org/TR/rdf12-concepts/#section-triple-terms
It is a triple in the role of an RDF term.

> Is a literal a term?

Yes

> Is a literal a node?

If it is in the graph

> Is a triple term a node?

If it is in the graph as a `<<( )>>` thingy.

>Is a triple term a term? (...If it isn't, why?)

Yes

Reply:
> If "foo" is not in the scope, then this rule applies

So scope is mapping a string, not a variable.

Reply:
Yes, it maps term to term, and I have taken the word "variable" out of the paragraph that started this thread.

Are you suggesting to rename shnex:var to something else?

Reply:
Why can't you say that the subject of `[ shnex:var "foo" ]` is (as in "describes", "represents" or "denotes", not in the RDF syntax where is a blank node) a <dfn>variable</dfn> with name "foo"?

After all:
```
shnex:var	sh:datatype xsd:string       The variable name, e.g. "focusNode".
            sh:minLength 1	        
```


Var expressions are expressions. They yield an RDF term when evaluated.
Here, they are the use of a variable in a role.

Reply:
I wanted to avoid the (formal) term variable because it is overloaded and in #650 you complain that there is no formal definition of variable. At this stage I am really not sure what you want here. Would you like to produce a PR or specific suggestion yourself instead of me guessing?

Reply:
Indeed #650 is a request have a definition or at least an anchor for "variable".

How can rules say "treat `[ shnex:var "foo" ]` as a (SPARQL) variable in filter expressions?
How can node expressions say "The variable name" in defining `shnex:var` without the domain being a variable?

Reply:
The anchor is for "var expression" https://w3c.github.io/data-shapes/shacl12-node-expr/#dfn-var-expression and you can then refer to the variable name as 

```
the <a>value</a> of <code>shnex:var</code> of the <a>var expression</a>.
```

this avoids me having to introduce a formal term for "variable" and other documents can freely redefine it.

If that's acceptable, please still approve this PR because it does some clean up.

Reply:
> this avoids me having to introduce a formal term for "variable" and other documents can freely redefine it.

Rules will now have to define evaluating node expressions because the input is a variable from earlier in the rule pattern.

Since we aren't defining a common evaluation model anymore, go ahead and merge this.
I'm not an editor of "Node Expression" so I will approve it this once.